File: frontend/src/data-services/api-services/models/queue-view-dto.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { LABEL, QUEUE_VIEW_TYPE } from '../../../constants';
import { QueueSortingSettingsDTO } from './queue-sorting-settings-dto';
import { QueueViewByTypeDto } from './queue-view-by-type-dto';
import { FilterConditionDto } from './settings/filter-condition-dto';

/**
 * QueueDTO model from API
 */
export interface QueueViewDTO {
    queueId: string;
    viewId: string;
    name: string;
    created: string;
    updated: string;
    size: number;
    allowedLabels: LABEL[];
    reviewers: string[];
    supervisors: string[];
    sorting: QueueSortingSettingsDTO;
    filters: FilterConditionDto[];

    /**
     * string($PnDTnHnMn.nS)
     * example: PT15M30S
     */
    processingDeadline: string;
    viewType: QUEUE_VIEW_TYPE;
    views: QueueViewByTypeDto[]

    /**
     * Determines if queue is a residual queue
     * The one that contains all order that did not apply to any created queue by their filters
     */
    residual: boolean;
}

File: backend/queues/src/main/java/com/griddynamics/msd365fp/manualreview/queues/model/dto/QueueViewDTO.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.griddynamics.msd365fp.manualreview.queues.model.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.griddynamics.msd365fp.manualreview.model.Label;
import com.griddynamics.msd365fp.manualreview.queues.model.ItemFilter;
import com.griddynamics.msd365fp.manualreview.queues.model.QueueSortSettings;
import com.griddynamics.msd365fp.manualreview.queues.model.QueueViewSettings;
import com.griddynamics.msd365fp.manualreview.queues.model.QueueViewType;
import com.griddynamics.msd365fp.manualreview.queues.validation.FieldConditionCombination;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Data;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.time.Duration;
import java.time.OffsetDateTime;
import java.util.Set;

import static com.griddynamics.msd365fp.manualreview.queues.model.Constants.SWAGGER_DURATION_EXAMPLE;
import static com.griddynamics.msd365fp.manualreview.queues.model.Constants.SWAGGER_DURATION_FORMAT;


/**
 * the DTO that reflects
 * {@link com.griddynamics.msd365fp.manualreview.queues.model.persistence.Queue}
 * object. For field description, please, see the original
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public class QueueViewDTO {
    @NotBlank
    private String queueId;
    @NotBlank
    private String viewId;
    @NotBlank
    private String name;
    @NotNull
    private OffsetDateTime created;
    private OffsetDateTime updated;
    @NotNull
    private Integer size;

    @NotNull
    private QueueViewType viewType;
    @NotNull
    private Set<Label> allowedLabels;
    private Set<QueueViewSettings> views;
    private Set<String> reviewers;
    @NotEmpty
    private Set<String> supervisors;
    @NotNull
    private QueueSortSettings sorting;
    private Set<@FieldConditionCombination @Valid ItemFilter> filters;
    @Schema(type = "string", format = SWAGGER_DURATION_FORMAT, example = SWAGGER_DURATION_EXAMPLE)
    @JsonFormat(shape = JsonFormat.Shape.STRING)
    private Duration processingDeadline;
    @NotNull
    private boolean residual;
    private boolean active;
}
